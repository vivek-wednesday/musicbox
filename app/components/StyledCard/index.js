/**
 *
 * StyledCard
 *
 */
import React, { useEffect, useRef, useState } from 'react';
import { Card, Progress, Col } from 'antd';
import {
  BackwardOutlined,
  PlayCircleOutlined,
  ForwardOutlined,
  CustomerServiceOutlined,
  PauseCircleOutlined
} from '@ant-design/icons';
import { isEmpty } from 'lodash';
import PropTypes from 'prop-types';
import { T } from '@components/T';
import { colors } from '@themes/index';

// Function to improve image resolution as well as deal with undefined data.
function getImage(str) {
  var escapedFind = '100x100'.replace(/([.*+?^=!:${}()|\]\\])/g, '\\$1');
  return str.replace(new RegExp(escapedFind, 'g'), '300x300');
}

function StyledCard(props) {
  const { musicData } = props;
  const { Meta } = Card;

  const [isPlaying, setIsPlaying] = useState(false);
  const [currentSong, setCurrentSong] = useState(null);
  const [progress, setProgress] = useState(0);
  const [itrvl, setItrvl] = useState(null);

  const audioList = useRef([]);

  useEffect(() => {
    clearInterval(itrvl);
    setCurrentSong(null);
    setProgress(0);
    setIsPlaying(false);
  }, [musicData]);

  function showProgress(e) {
    setItrvl(setInterval(() => setProgress((e.currentTime / e.duration) * 100), 1000));
  }

  function controlAudio(val, type) {
    const ele = audioList.current[val];
    if (type === 'backward') {
      ele.currentTime -= 5;
    } else if (type === 'forward') {
      ele.currentTime += 5;
    } else if (type === 'play') {
      ele.play();
      setIsPlaying(true);
      setCurrentSong(val);
      showProgress(ele);
    } else if (type === 'pause') {
      ele.pause();
      setIsPlaying(false);
      clearInterval(itrvl);
    }
  }

  function showProgressBar(id) {
    if (currentSong === id) {
      return <Progress percent={progress} showInfo={false} />;
    }
  }

  return musicData === undefined ? (
    <T data-testid="no-music-data" id="no_results" />
  ) : isEmpty(musicData.results) ? (
    <T data-testid="no-music-data" id="no_results" />
  ) : (
    musicData.results.slice(0, 10).map((item, index) => {
      const data = musicData.results[index];
      return (
        <Col key={index} >
          <Card
            data-testid="card-wrapper"
            key={data.trackId}
            className="neumorphic"
            style={{ width: 300 }}
            cover={
              <img
                alt="example"
                src={getImage(data.artworkUrl100)}
                style={{ borderBottomLeftRadius: 35, borderBottomRightRadius: 35, padding: 10 }}
              />
            }
            actions={[
              <BackwardOutlined
                key="backward"
                style={{ fontSize: 35, color: colors.secondaryBtn }}
                onClick={() => {
                  controlAudio(index, 'backward');
                }}
              />,
              <>
                {isPlaying ? (
                  currentSong === index ? (
                    <PauseCircleOutlined
                      key="play"
                      style={{ fontSize: 40, color: colors.primaryBtn }}
                      onClick={() => {
                        controlAudio(index, 'pause');
                      }}
                    />
                  ) : (
                    <PlayCircleOutlined key="play" style={{ fontSize: 40, color: colors.secondaryBtn }} />
                  )
                ) : (
                  <PlayCircleOutlined
                    key="play"
                    style={{ fontSize: 40, color: colors.primaryBtn }}
                    onClick={() => {
                      controlAudio(index, 'play');
                    }}
                  />
                )}
              </>,
              <ForwardOutlined
                key="forward"
                style={{ fontSize: 35, color: colors.secondaryBtn }}
                onClick={() => {
                  controlAudio(index, 'forward');
                }}
              />
            ]}
          >
            <Meta
              avatar={<CustomerServiceOutlined style={{ fontSize: 20 }} spin={currentSong === index ? true : false} />}
              title={data.trackCensoredName}
            />
            <Meta title={data.artistName} description={showProgressBar(index)} />
            <audio ref={ele => audioList.current[index] = ele} src={data.previewUrl} preload="none" loop />
          </Card>
        </Col>
      );
    })
  );
}

StyledCard.propTypes = {
  musicData: PropTypes.object
};

export default StyledCard;
