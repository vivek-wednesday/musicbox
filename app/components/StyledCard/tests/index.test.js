/**
 *
 * Tests for StyledCard
 *
 */

import React from 'react';
import { renderWithIntl } from '@utils/testUtils';
import StyledCard from '../index';
import { data } from './testData';
import { fireEvent} from "@testing-library/react"
import {BrowserRouter} from "react-router-dom"

describe('<StyledCard />', () => {

  const sampleData = {
    results: data
  }

  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<BrowserRouter><StyledCard musicData={sampleData} /></BrowserRouter>);
    expect(baseElement).toMatchSnapshot();
  });
  
  it('should render cards with music data', () => {
    const { getAllByTestId } = renderWithIntl(<BrowserRouter><StyledCard musicData={sampleData} /></BrowserRouter>);
    expect(getAllByTestId('card-wrapper')).toBeTruthy;
  });

  it('should render play button', () => {
    const {getAllByTestId} = renderWithIntl(<BrowserRouter><StyledCard musicData={sampleData} /></BrowserRouter>);
    expect(getAllByTestId('play')).toBeTruthy;
  });

  it('should render play button and click it', async () => {
    const {getAllByTestId} = renderWithIntl(<BrowserRouter><StyledCard musicData={sampleData} /></BrowserRouter>);
    expect(fireEvent(
      getAllByTestId('play')[0],
      new MouseEvent('click'),
    )).toBeTruthy;
  });

  it('should not render any pause button', () => {
    const {getAllByTestId} = renderWithIntl(<BrowserRouter><StyledCard musicData={sampleData} /></BrowserRouter>);
    expect(getAllByTestId('play')).toBeFalsy;
  });

  it('should render forward button', () => {
    const {getAllByTestId} = renderWithIntl(<BrowserRouter><StyledCard musicData={sampleData} /></BrowserRouter>);
    expect(getAllByTestId('forward')).toBeTruthy;
  });

  it('should render backward button', () => {
    const {getAllByTestId} = renderWithIntl(<BrowserRouter><StyledCard musicData={sampleData} /></BrowserRouter>);
    expect(getAllByTestId('backward')).toBeTruthy;
  });

  it('should render error message without music data', () => {
    sampleData.results.length=0;
    const { getAllByTestId } = renderWithIntl(<StyledCard musicData={sampleData} />);
    expect(getAllByTestId('no-music-data')).toBeTruthy;
  });

});
