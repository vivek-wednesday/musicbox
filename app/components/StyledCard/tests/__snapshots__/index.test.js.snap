// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StyledCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <p
      class="T__StyledText-gjlic1-0 kYcyRq"
      data-testid="no-music-data"
    >
      Oops! There's nothing here.
    </p>
  </div>
</body>
`;
